libmeepgeom/dft-fields.cpp

```diff
@@ -0,0 +1,222 @@
+
+  // read dataset 1
+  h5file f1(file1, h5file::READONLY, false);
+  int rank1, *dims1=new int[expected_rank];
+  snprintf(dataname,100,"%s.r",name1);
+  double *rdata1 = f1.read(dataname, &rank1, dims1, expected_rank);
+  snprintf(dataname,100,"%s.i",name1);
+  double *idata1 = f1.read(dataname, &rank1, dims1, expected_rank);
+  if (!rdata1 || !idata1) return -1.0;
+
+  // read dataset 2
+  h5file f2(file2, h5file::READONLY, false);
+  int rank2, *dims2=new int[expected_rank];
+  snprintf(dataname,100,"%s.r",name2);
+  double *rdata2 = f2.read(dataname, &rank2, dims2, expected_rank);
+  snprintf(dataname,100,"%s.i",name2);
+  double *idata2 = f2.read(dataname, &rank2, dims2, expected_rank);
+  if (!rdata2 || !idata2) return -1.0;
+
+  // check same size
+  bool same_size = (rank1==rank2);
+  for(int d=0; same_size && d<rank1; d++)
+   if(dims1[d]!=dims2[d])
+    same_size = false;
+  if (!same_size)
+   meep::abort("data size mismatch\n");
+
+  // first pass to normalize each dataset to its maximum absolute magnitude;
+  // we also note the phase difference between the datasets at their points
+  // of maximum magnitude so we can compensate for this in the comparison below.
+  int length = dims1[0];
+  for(int d=1; d<rank1; d++)
+   length*=dims1[d];
+
+  double max_abs1=0.0, max_abs2=0.0;
+  double max_arg1=0.0, max_arg2=0.0;
+  for(int n=0; n<length; n++)
+   { cdouble z1 = cdouble(rdata1[n], idata1[n]);
+     if (abs(z1) > max_abs1)
+      { max_abs1 = abs(z1);
+        max_arg1 = arg(z1);
+      }
+     cdouble z2 = cdouble(rdata2[n], idata2[n]);
+     if (abs(z2) > max_abs2)
+      { max_abs2 = abs(z2);
+        max_arg2 = arg(z2);
+      }
+   }
+
+  // second pass to get L2 norm of difference between normalized data sets
+  double norm1=0.0, norm2=0.0, normdiff=0.0;
+  cdouble phase_fac = exp( cdouble(0,1)*(max_arg1 - max_arg2) );
+  for(int n=0; n<length; n++)
+   { cdouble z1 = cdouble(rdata1[n], idata1[n]) / max_abs1;
+     cdouble z2 = phase_fac*cdouble(rdata2[n], idata2[n]) / max_abs2;
+     norm1    += norm(z1);
+     norm2    += norm(z2);
+     normdiff += norm(z1-z2);
+   }
+  norm1    = sqrt(norm1)    / ((double)length);
+  norm2    = sqrt(norm2)    / ((double)length);
+  normdiff = sqrt(normdiff) / ((double)length);
+
+  if (am_master())
+   { FILE *ff=fopen("/tmp/dft-fields.log","a");
+     fprintf(ff,"length %i:\n",length);
+     fprintf(ff,"norm1    : %e\n",norm1);
+     fprintf(ff,"norm2    : %e\n",norm2);
+     fprintf(ff,"normdiff : %e\n",normdiff);
+     fprintf(ff,"\n");
+     fclose(ff);
+   }
+
+  return normdiff / fmax(norm1,norm2);
+}
+
+
+/***************************************************************/
+/***************************************************************/
+/***************************************************************/
+int main(int argc, char *argv[])
+{
+  initialize mpi(argc, argv);
+
+#define NUM_RESOLUTIONS 3    
+  double resolutions[NUM_RESOLUTIONS]={10.0, 20.0, 40.0};
```

Comment:
Since the range of resolutions up to 40 makes this test too time intensive and thus impractical to be included in the test suite, can the same trend of decreasing error with resolution be demonstrated with a smaller range (i.e., 8, 12, 16) which executes much faster?